src/components/TodoList/TodoList.tsx

```diff
@@ -1,18 +1,59 @@
-import React from 'react';
+import React, { useEffect, useMemo, useState } from 'react';
 import classNames from 'classnames';
 import { useDispatch } from 'react-redux';
 import { useAppSelector } from '../../app/hooks';
 import { actions as currentTodoActions } from '../../features/currentTodo';
+import { actions as todosActions } from '../../features/todos';
 import { Todo } from '../../types/Todo';
+import { Filter } from '../../types/Filter';
+import { getTodos } from '../../api';
+import { Loader } from '../Loader';
 
-type State = {
-  todos: Todo[];
-};
-
-export const TodoList: React.FC<State> = ({ todos }) => {
+export const TodoList: React.FC = () => {
   const currentTodo = useAppSelector(state => state.currentTodo);
+  const todos = useAppSelector(state => state.todos);
+  const { query, filter } = useAppSelector(state => state.filter);
+
   const dispatch = useDispatch();
 
+  const [isLoading, setIsLoading] = useState(true);
+
+  useEffect(() => {
+    getTodos()
+      .then((todosFromApi: Todo[]) =>
+        dispatch(todosActions.setTodos(todosFromApi)),
+      )
+      .catch(() => {
+        alert("Couldn't fetch the todos");
+      })
+      .finally(() => setIsLoading(false));
+  }, [dispatch]);
+
+  const displayedTodos = useMemo(() => {
+    let todosCopy = [...todos];
+
+    switch (filter) {
+      case Filter.All:
+        break;
+      case Filter.Active:
+        todosCopy = todosCopy.filter((todo: Todo) => todo.completed === false);
+        break;
+      case Filter.Completed:
+        todosCopy = todosCopy.filter((todo: Todo) => todo.completed === true);
+        break;
+    }
+
+    todosCopy = todosCopy.filter((todo: Todo) =>
+      todo.title.toLocaleLowerCase().includes(query.toLocaleLowerCase()),
+    );
+
+    return todosCopy;
+  }, [filter, query, todos]);
+
+  if (isLoading) {
+    return <Loader />;
+  }
+
   if (!todos.length) {
```

Comment:
you should use displayedTodos
![image](https://github.com/mate-academy/react_redux-list-of-todos/assets/126968958/1dbbe2ef-77c1-43f0-95eb-9d0e3c7c65ca)